### PR TITLE
Faster joins: don't stall when a user joins during a fast join

### LINKED ISSUE #12801: Faster joins: Handle a second join while syncing state for the first one
We want to avoid launching a second partial state syncing job for the same room.

### PR DESCRIPTION
**Pull request by MatMaul**
Fixes #12801.
Complement tests are at
https://github.com/matrix-org/complement/pull/567.

Avoid blocking on full state when handling a subsequent join into a
partial state room.

Also always perform a remote join into partial state rooms, since we do
not know whether the joining user has been banned and want to avoid
leaking history to banned users.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by MatMaul** [comment]
We had a discussion about `do_invite_join` [here](https://matrix.to/#/!vcyiEtMVHIhWXcJAfl:sw1v.org/$rN2eFSTfP5CasVf08fkhW1xV5pdpuZBM8H7Ynmx1mOc?via=matrix.org&via=element.io&via=envs.net).
TLDR:
1) this method currently removes current room extremities as a first step. If the second `send_join` is not accepted by the resident server, we don't get any new extremity and we can't send messages anymore.
2) un-partial-stating during this is not handled and should fail

For 1 we could store the membership event received from `send_join` (if succeeded) in the same transaction as deleting the old extremities.

**Comment by squahtx** [comment]
One of the ways an un-partial-stating during join could manifest is in a `PartialStateConflictError` (note the todo).
https://github.com/matrix-org/synapse/blob/54c012c5a8722725cf104fa6205f253b5b9b0192/synapse/handlers/federation.py#L643-L665

We also need to be careful that we don't return a room to partial state, because that breaks a lot of assumptions.

**Comment by MatMaul** [comment]
> 1) this method currently removes current room extremities as a first step. If the second send_join is not accepted by the resident server, we don't get any new extremity and we can't send messages anymore.
> For 1 we could store the membership event received from `send_join` (if succeeded) in the same transaction as deleting the old extremities.

It's not possible since storing events involves potentially doing so in another worker.

What I've done instead to mitigate the issue:
- do not remove forward extremities if the room is partially joined (in SQL to avoid races)
- move `_clean_room_for_join` call after `send_join`, so we avoid cleaning up if the join is failing anyway

**Comment by MatMaul** [comment]
I've added a way to fallback to local join if a room is unpartial stated during remote join.

However I am not sure this is really useful: retrying the remote join is a lot simpler, and I think that should be enough ?

Actually it's possible we will never hit the case: the event context is calculated by another resident server in a `make_event` remote call. It can contain local state (if a message was send to the resident server from the joining-already-joined server in the meantime), but unsure if/how this plays out with the unpartial state.

I think we should just retry the whole `make/send_join` in case of and that should be enough, in the same principle as 54c012c5a8722725cf104fa6205f253b5b9b0192.

**Comment by MatMaul** [comment]
> > However I am not sure this is really useful: retrying the remote join is a lot simpler, and I think that should be enough ?
> 
> Either approach (local or remote) sounds fine to me. Since this is already a race, it doesn't really matter which outcome you go for — so quite happy for you to choose the most simple and well-reasoned approach.

I tried the retry remote join instead, and it seems to end up with even more subtle changes because of the `else` and `finally` clause of the impacted `try`. I decide to keep it the way it is now (fallback to local join).

File: synapse/handlers/federation.py [hunk 1]
```diff
@@ -629,7 +633,8 @@ async def do_invite_join(
                 state_events=state,
             )
 
-            if ret.partial_state:
+            already_partially_joined = await self.store.is_partial_state_room(room_id)
```
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/handlers/federation.py:636]: There's still a race:
1. We enter `do_invite_join()`.
2. The room is un-partial stated in the background.
3. We read the partial state flag and see that it's not present.
4. We then proceed to re-partial state the room (!) which breaks a key assumption.

I propose we introduce a lock that guarantees that no other code may toggle the partial stated-ness of the room while the lock is held. That is, when joining:
1. We take the lock for the room
2. We do a partial send-join when i) we aren't in the room or ii) the room is partial stated. Corollary: we do a full join when we are in the room and it's not partial stated. This is the same condition as mentioned in another comment.
3. We persist the join, which should no longer be able to raise a `PartialStateConflictError`
4. We release the lock right after the code that kicks off the sync.

When syncing, we take the lock when we're about to finish and un-partial state the room.

We can use https://github.com/matrix-org/synapse/blob/develop/synapse/storage/databases/main/lock.py for the locking, but it needs an `acquire()` method to be implemented (keep attempting `try_acquire()` and sleeping? not ideal but implementing it properly with replication is a fair bit of work)

TODO: check in #synapse-devs if this makes sense.

File: synapse/handlers/federation.py [hunk 2]
```diff
@@ -654,15 +659,14 @@ async def do_invite_join(
...
-                # a 429, as a hint to the client to try again.
+                # stated. We ought to be doing a local join instead.
                 # TODO(faster_joins): `_should_perform_remote_join` suggests that we may
                 #   do a remote join for restricted rooms even if we have full state.
                 logger.error(
                     "Room %s was un-partial stated while processing remote join.",
                     room_id,
                 )
-                raise LimitExceededError(msg=e.msg, errcode=e.errcode, retry_after_ms=0)
+                raise e
```
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/handlers/federation.py:669]: The proposal with a lock in another comment should eliminate `PartialStateConflictError`s and make this comment moot.

File: synapse/handlers/federation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/handlers/federation.py:636]: @DMRobertson Heads up that some of the above may impact leaves while partial stated. I'm not entirely sure yet.

**Review by squahtx — COMMENTED** [review]
I did some thinking about the races and made some proposals for addressing them. I reckon introducing a lock around un-partial stating of rooms will fix them, though it's worth discussing it in #synapse-devs first. Sorry about the PR churn.

File: synapse/handlers/event_auth.py [hunk 3]
```diff
@@ -214,15 +214,13 @@ async def check_restricted_join_rules(
...
+            prev_membership: The current membership state for this user.
 
         Raises:
             AuthError if the user cannot join the room.
         """
         # If the member is invited or currently joined, then nothing to do.
-        if prev_member_event and (
-            prev_member_event.membership in (Membership.JOIN, Membership.INVITE)
-        ):
+        if prev_membership in (Membership.JOIN, Membership.INVITE):
```
Comment (DMRobertson) [inline synapse/handlers/event_auth.py:223]: I can see that we're only using the membership in this function; this looks correct. Any particular reason for making this change though?

If we're making these changes, could we change the docstring to explain what `prev_membership=None` means? I would guess this means that they've never been a member of the room in question?

File: synapse/handlers/federation.py [hunk 4]
```diff
@@ -562,6 +562,10 @@ async def do_invite_join(
             joinee: The User ID of the joining user.
 
             content: The event content to use for the join event.
+
+        Raises:
+            PartialStateConflictError if homeserver was already in the room and it is
+                no longer partial stated.
```
Comment (DMRobertson) [inline synapse/handlers/federation.py:568]: I can see this is taken from the comment below, but it hard to understand when this occurs. Is the point that

1. HS is partially joined to the room
2. This function gets called
3. We wait for a response from the remote HS
4. The resync for this room completes; the room is now no longer partially stated
5. We get a response from (3); but we should discard it and create a local join ourselves.

File: synapse/handlers/federation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (DMRobertson) [reply synapse/handlers/federation.py:636]: > Maybe the partial state flag should imply that we are currently trying to sync partial state

FWIW this was my internal model for the meaning of that flag.

I'm not quite following all the moving changes here. I might have to pick your brains tomorrow

File: synapse/handlers/room_member.py [hunk 5]
```diff
@@ -823,8 +841,14 @@ async def update_membership_locked(
 
         latest_event_ids = await self.store.get_prev_events_for_room(room_id)
 
+        # There are 2 reasons that mandate to have the full state:
+        # - to calculate `is_host_in_room`, however we now approximate that we consider the server
+        # in the room if we are still in the middle of a fast join.
+        # - to calculate if another user trying to join is allowed to. In this case we now
+        # make another `make/send_join` to a resident server to validate that. We could just
+        # accept it but it would mean that we would potentially leak the history to a banned user.
```
Comment (DMRobertson) [inline synapse/handlers/room_member.py:849]: Not fully sure what you're getting at here. Are you trying to enumerate the ways in which we use `state_before_join` and argue that it's okay that if this is only partial state?

File: synapse/handlers/room_member.py [hunk 6]
```diff
@@ -1634,7 +1677,13 @@ async def _make_and_store_3pid_invite(
...
         return event, stream_id
 
-    async def _is_host_in_room(self, current_state_ids: StateMap[str]) -> bool:
+    async def _is_host_in_room(
+        self, is_partial_state_room: bool, current_state_ids: StateMap[str]
+    ) -> bool:
+        # When we only have a partial state, let's assume we are still in the room
+        if is_partial_state_room:
+            return True
+
```
Comment (DMRobertson) [inline synapse/handlers/room_member.py:1686]: I think this means that #12802 is going to have to ensure this returns False as soon as the last person has left? Feels like there's race potential

File: synapse/handlers/federation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/handlers/federation.py:636]: > Maybe the partial state flag should imply that we are currently trying to sync partial state

We now think that the partial state flag just means that there are partial state events in the room and does not imply that there is a partial state sync running (because the partial state sync may have failed). It also does not imply that the local user currently has a member in the room.

**Comment by squahtx** [comment]
I tried to address the issues with restricted join in #14882, but it seems to have ended up second join-proofing `do_invite_join` and replacing that part of this PR. Please ignore all my comments about that method if #14882 lands.

File: synapse/handlers/federation.py (hunk 4 above)
Comment (squahtx) [reply synapse/handlers/federation.py:575]: That's right (or was right prior to #14882, which guarantees no `PartialStateConflictError`s).

File: synapse/handlers/federation.py [hunk 7]
```diff
@@ -614,6 +616,8 @@ async def do_invite_join(
 
             logger.debug("do_invite_join event: %s", event)
 
+            await self._clean_room_for_join(room_id)
```
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/handlers/federation.py]: Done as part of#14882.

File: synapse/handlers/federation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/handlers/federation.py]: The proposed locking was implemented in #14882.

File: synapse/handlers/federation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/handlers/federation.py]: #14882 implements the proposed lock, so we can no longer observe a `PartialStateConflictError`.

File: synapse/storage/databases/main/event_federation.py [hunk 8]
```diff
@@ -2016,7 +2016,12 @@ async def clean_room_for_join(self, room_id: str) -> None:
         )
 
     def _clean_room_for_join_txn(self, txn: LoggingTransaction, room_id: str) -> None:
-        query = "DELETE FROM event_forward_extremities WHERE room_id = ?"
+        query = """
+                DELETE FROM event_forward_extremities
+                    WHERE room_id = ? AND room_id NOT IN (
+                        SELECT room_id FROM partial_state_rooms
```
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/storage/databases/main/event_federation.py]: #14882 implements that membership check proposed in the other review comment, which replaces this filter.

**Comment by squahtx** [comment]
Now that #14882 has landed, I've gone and resolved the merge conflicts.

All the PR comments plus changes for `do_invite_join` have gone into #14882, so that method is now compatible with second joins.
Also `do_invite_join` will no longer raise a PartialStateConflictError because there is locking now.

File: synapse/handlers/room_member.py [hunk 9]
```diff
@@ -881,11 +902,11 @@ async def update_membership_locked(
             if action == "kick":
                 raise AuthError(403, "The target user is not in the room")
 
-        is_host_in_room = await self._is_host_in_room(state_before_join)
+        is_host_in_room = await self._is_host_in_room(partial_state_before_join)
```
Comment (squahtx) [inline synapse/handlers/room_member.py:905]: I think knocking (line +1052) is going to be broken in that we can end up sending a knock event from a user with a ban we don't know about.
We can't exactly await full state, since we're holding the room membership lock. Any thoughts on what we should do in this case?

**Review by DMRobertson — COMMENTED** [review]
I think this LGTM and we should land it, provided that we have reasonable complement coverage to land with it.

File: synapse/handlers/room_member.py (hunk 5 above)
Comment (DMRobertson) [reply synapse/handlers/room_member.py]: Many thanks, this is much appreciated.

File: synapse/handlers/room_member.py (hunk 6 above)
Comment (DMRobertson) [reply synapse/handlers/room_member.py]: I think(?) you've addressed this by making it the caller's responsibility to interpret the return value of this function.

**Comment by DMRobertson** [comment]
Also: is the description of this PR still accurate to what it does, or should we update it?

File: synapse/handlers/room_member.py (hunk 9 above)
Comment (squahtx) [reply synapse/handlers/room_member.py:905]: It's a bit different. Previously, we would do a remote /make_knock+/send_knock and the remote homeserver would tell us to go away we were banned.
Now, since we are in the room but do not have full state, we can add a local knock into the DAG that would otherwise be rejected and cite it as a prev/auth event.

File: synapse/handlers/federation_event.py [hunk 10]
```diff
@@ -524,11 +525,57 @@ async def process_remote_join(
             "Peristing join-via-remote %s (partial_state: %s)", event, partial_state
         )
         with nested_logging_context(suffix=event.event_id):
+            if partial_state:
+                # When handling a second partial state join into a partial state room,
+                # the returned state will exclude the membership from the first join. To
+                # preserve prior memberships, we try to compute the state before the
```
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/handlers/federation_event.py]: Updated the comment

File: synapse/handlers/federation_event.py [hunk 11]
```diff
@@ -524,11 +525,57 @@ async def process_remote_join(
...
+                            room_id, seen_event_ids, await_full_state=False
+                        )
+                    ).values()
+                )
+
+                # When there are prev events we do not have the state for, we state
+                # resolve with the state returned by the remote homeserver.
+                if missing_event_ids or len(state_maps_to_resolve) == 0:
+                    state_maps_to_resolve.append(
+                        {(e.type, e.state_key): e.event_id for e in state}
```
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/handlers/federation_event.py:559]: There doesn't appear to be such a check anywhere.

File: synapse/handlers/room_member.py [hunk 12]
```diff
@@ -1109,28 +1135,43 @@ async def _should_perform_remote_join(
...
+            previous_membership = prev_member_event.membership
+
+        # If we are not fully joined yet, and the target is not already in the room,
+        # let's do a remote join so another server with the full state can validate
+        # that the user has not been banned for example.
+        # We could just accept the join and wait for state res to resolve that later on
+        # but we would then leak room history to this person until then, which is pretty
+        # bad.
+        if is_partial_state_room and previous_membership != Membership.JOIN:
+            return True, remote_room_hosts
```
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/handlers/room_member.py:1153]: Added a test as part of https://github.com/matrix-org/complement/pull/567. Synapse develop currently blocks and fails the test.

**Comment by squahtx** [comment]
> Also: is the description of this PR still accurate to what it does, or should we update it?

Er, let's update it.

File: synapse/handlers/room_member.py (hunk 9 above)
Comment (DMRobertson) [reply synapse/handlers/room_member.py:905]: Hmm. Would it be worth requiring knocks to go over federation during the resync? (Not necessarily in this PR)

Comment (squahtx) [reply synapse/handlers/room_member.py:905]: I'm not wholly convinced the remote knock code is suitable. It persists the knock event as an outlier. I'm not sure what happens if we then receive the knock event over federation.

Comment (DMRobertson) [reply synapse/handlers/room_member.py:905]: I'm inclined to land this as-is and kick any knocking cans down the road.

**Review by DMRobertson — APPROVED** [review]
